lib/srv/discovery/status.go

```diff
@@ -0,0 +1,167 @@
+	// Allowed values are:
+	// - DISCOVERY_CONFIG_STATE_SYNCING
+	// - DISCOVERY_CONFIG_STATE_ERROR
+	// - DISCOVERY_CONFIG_STATE_RUNNING
+	state               string
+	errorMessage        *string
+	lastSyncTime        time.Time
+	discoveredResources uint64
+}
+
+func (d *awsSyncStatus) iterationFinished(fetchers []aws_sync.AWSSync, pushErr error, lastUpdate time.Time) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.awsSyncStatus = make(map[string][]awsSyncResult)
+	for _, fetcher := range fetchers {
+		// Only update the status for fetchers that are from the discovery config.
+		if !fetcher.IsFromDiscoveryConfig() {
+			continue
+		}
+
+		count, statusErr := fetcher.Status()
+		statusAndPushErr := trace.NewAggregate(statusErr, pushErr)
+
+		fetcherResult := awsSyncResult{
+			state:               discoveryconfigv1.DiscoveryConfigState_DISCOVERY_CONFIG_STATE_RUNNING.String(),
+			lastSyncTime:        lastUpdate,
+			discoveredResources: count,
+		}
+
+		if statusAndPushErr != nil {
+			errorMessage := statusAndPushErr.Error()
+			fetcherResult.errorMessage = &errorMessage
+			fetcherResult.state = discoveryconfigv1.DiscoveryConfigState_DISCOVERY_CONFIG_STATE_ERROR.String()
+		}
+
+		d.awsSyncStatus[fetcher.DiscoveryConfigName()] = append(d.awsSyncStatus[fetcher.DiscoveryConfigName()], fetcherResult)
+	}
+}
+
+func (d *awsSyncStatus) discoveryConfigs() []string {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	ret := make([]string, 0, len(d.awsSyncStatus))
+	for k := range d.awsSyncStatus {
+		ret = append(ret, k)
+	}
+	return ret
+}
+
+func (d *awsSyncStatus) iterationStarted(fetchers []aws_sync.AWSSync, lastUpdate time.Time) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.awsSyncStatus = make(map[string][]awsSyncResult)
+	for _, fetcher := range fetchers {
+		// Only update the status for fetchers that are from the discovery config.
+		if !fetcher.IsFromDiscoveryConfig() {
+			continue
+		}
+
+		fetcherResult := awsSyncResult{
+			state:        discoveryconfigv1.DiscoveryConfigState_DISCOVERY_CONFIG_STATE_SYNCING.String(),
+			lastSyncTime: lastUpdate,
+		}
+
+		d.awsSyncStatus[fetcher.DiscoveryConfigName()] = append(d.awsSyncStatus[fetcher.DiscoveryConfigName()], fetcherResult)
+	}
+}
+
+func (d *awsSyncStatus) mergeIntoGlobalStatus(discoveryConfigName string, existingStatus discoveryconfig.Status) discoveryconfig.Status {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+
+	awsStatusFetchers, found := d.awsSyncStatus[discoveryConfigName]
+	if !found {
+		return existingStatus
+	}
+
+	for _, fetcher := range awsStatusFetchers {
+		existingStatus.DiscoveredResources = existingStatus.DiscoveredResources + fetcher.discoveredResources
+
+		// Each DiscoveryConfigStatus has a global State and Error Message, but those are produced per Fetcher.
+		// We choose to keep the most informative states by favoring error states/messages.
+		if existingStatus.ErrorMessage == nil || *existingStatus.ErrorMessage == "" {
+			existingStatus.ErrorMessage = fetcher.errorMessage
+		}
```

Comment:
Can we aggregate the error messages in case there are multiple erroring fetchers for the same discovery config?

Reply:
Yes, that sounds good 👍